File: solutions/typescript/dnd-character/1/dnd-character.ts
export class DnDCharacter {
  public strength = 0;
  public dexterity = 0;
  public constitution = 0;
  public intelligence = 0;
  public wisdom = 0;
  public charisma = 0;
  public hitpoints = 10;

  constructor() {
    this.strength = DnDCharacter.getAbilityScore();
    this.dexterity = DnDCharacter.getAbilityScore();
    this.constitution = DnDCharacter.getAbilityScore();
    this.intelligence = DnDCharacter.getAbilityScore();
    this.wisdom = DnDCharacter.getAbilityScore();
    this.charisma = DnDCharacter.getAbilityScore();
    this.hitpoints += DnDCharacter.getModifierFor(this.constitution);
  }

  public static generateAbilityScore(): number {
    return this.getAbilityScore();
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2);
  }

  private static getAbilityScore(): number {
    let min = Number.MAX_VALUE;
    let sum = 0;
    for (let i = 0; i < 4; i++) {
      const rollNum = this.getDiceRollNum();
      if (rollNum < min)
        min = rollNum
      sum += rollNum;
    }
    sum -= min;
    return sum;
  }

  private static getDiceRollNum(): number {
    return Math.floor(Math.random() * 7);
  }
}

File: solutions/csharp/dnd-character/1/DndCharacter.cs
using System;

public class DndCharacter
{
    private static Random random = new Random();
    public int Strength { get; init; }
    public int Dexterity { get; init; }
    public int Constitution { get; init; }
    public int Intelligence { get; init; }
    public int Wisdom { get; init; }
    public int Charisma { get; init; }
    public int Hitpoints { get => Modifier(Constitution) + 10; }

    public static int Modifier(int score) => (int)Math.Floor((score - 10) / 2.0);

    public static int Ability()
    {
        int minScore = int.MaxValue;
        int sum = 0;
        for (int i = 0; i < 4; i++)
        {
            int randomScore = RollDie();
            minScore = randomScore < minScore ? randomScore : minScore;
            sum += randomScore;
        }
        sum -= minScore;
        return sum;
    }

    public static DndCharacter Generate()
    {
        return new DndCharacter()
        {
            Strength = Ability(),
            Dexterity = Ability(),
            Constitution = Ability(),
            Intelligence = Ability(),
            Wisdom = Ability(),
            Charisma = Ability(),
        };
    }

    private static int RollDie() => random.Next(1, 7);
}
